File: drill_optimizer/python/drill_optimizer.py
"""PCB Drill Path Optimizer.

Optimizes the order of drilling holes to minimize total travel distance.
Uses Nearest Neighbor construction + 2-opt local search improvement.
Groups holes by tool to minimize tool changes.
"""

from __future__ import annotations

import math
import time
from dataclasses import dataclass, field
from itertools import permutations

from excellon_parser import DrillHole, DrillProject, ToolDefinition


@dataclass
class OptimizationResult:
    ordered_holes: list[DrillHole]
    total_distance: float
    tool_changes: int
    original_distance: float
    improvement: float
    tool_stats: list[dict]


def _dist(a: DrillHole | tuple, b: DrillHole | tuple) -> float:
    if isinstance(a, DrillHole):
        ax, ay = a.x, a.y
    else:
        ax, ay = a
    if isinstance(b, DrillHole):
        bx, by = b.x, b.y
    else:
        bx, by = b
    return math.sqrt((ax - bx) ** 2 + (ay - by) ** 2)


def _path_distance(holes: list[DrillHole]) -> float:
    total = 0.0
    current = (0.0, 0.0)
    for h in holes:
        total += _dist(current, h)
        current = (h.x, h.y)
    return total


def nearest_neighbor(points: list[tuple[float, float]], start: tuple[float, float] = (0, 0)) -> list[int]:
    """Greedy TSP: always move to the nearest unvisited point."""
    n = len(points)
    if n <= 1:
        return list(range(n))

    visited = [False] * n
    route = []
    current = start

    for _ in range(n):
        best_dist = float("inf")
        best_idx = -1
        for j in range(n):
            if not visited[j]:
                d = _dist(current, points[j])
                if d < best_dist:
                    best_dist = d
                    best_idx = j
        visited[best_idx] = True
        route.append(best_idx)
        current = points[best_idx]

    return route


def two_opt_improve(
    points: list[tuple[float, float]],
    route: list[int],
    max_iterations: int = 200,
    time_limit: float = 5.0,
) -> list[int]:
    """2-opt local search to improve a route."""
    n = len(route)
    if n < 4:
        return list(route)

    improved = list(route)
    start_time = time.monotonic()
    iterations = 0

    did_improve = True
    while did_improve and iterations < max_iterations:
        did_improve = False
        iterations += 1

        if time.monotonic() - start_time > time_limit:
            break

        for i in range(1, n - 2):
            for j in range(i + 2, n):
                a, b = improved[i - 1], improved[i]
                c, d = improved[j], improved[j + 1] if j + 1 < n else improved[j]

                current_d = _dist(points[a], points[b]) + _dist(points[c], points[d])
                new_d = _dist(points[a], points[c]) + _dist(points[b], points[d])

                if new_d < current_d - 1e-10:
                    improved[i : j + 1] = reversed(improved[i : j + 1])
                    did_improve = True

    return improved


def optimize(project: DrillProject, max_iterations: int = 200, time_limit: float = 5.0) -> OptimizationResult:
    """Optimize drilling path for a complete project."""
    if not project.holes:
        return OptimizationResult(
            ordered_holes=[], total_distance=0, tool_changes=0,
            original_distance=0, improvement=0, tool_stats=[],
        )

    original_distance = _path_distance(project.holes)

    # Group by tool
    tool_groups: dict[int, list[DrillHole]] = {}
    for hole in project.holes:
        tool_groups.setdefault(hole.tool, []).append(hole)

    tool_numbers = sorted(tool_groups.keys())

    # Optimize each group
    optimized_groups: dict[int, list[DrillHole]] = {}
    for tn in tool_numbers:
        holes = tool_groups[tn]
        points = [(h.x, h.y) for h in holes]
        nn_route = nearest_neighbor(points, (0, 0))
        opt_route = two_opt_improve(points, nn_route, max_iterations, time_limit)
        optimized_groups[tn] = [holes[i] for i in opt_route]

    # Order tool groups
    group_order = _order_groups(optimized_groups, tool_numbers)

    # Build final list
    ordered_holes: list[DrillHole] = []
    changes = 0
    current_tool = None
    for tn in group_order:
        if current_tool is not None and current_tool != tn:
            changes += 1
        current_tool = tn
        ordered_holes.extend(optimized_groups[tn])

    total_distance = _path_distance(ordered_holes)
    improvement = (original_distance - total_distance) / original_distance * 100 if original_distance > 0 else 0

    tool_stats = []
    for tn in tool_numbers:
        holes = optimized_groups[tn]
        td = project.tools.get(tn)
        tool_stats.append({
            "tool_number": tn,
            "diameter": td.diameter if td else 0,
            "hit_count": len(holes),
            "sub_path_distance": _path_distance(holes),
        })

    return OptimizationResult(
        ordered_holes=ordered_holes,
        total_distance=total_distance,
        tool_changes=changes,
        original_distance=original_distance,
        improvement=improvement,
        tool_stats=tool_stats,
    )


def _order_groups(groups: dict[int, list[DrillHole]], tool_numbers: list[int]) -> list[int]:
    """Order tool groups to minimize inter-group travel."""
    if len(tool_numbers) <= 1:
        return tool_numbers

    group_info = []
    for tn in tool_numbers:
        holes = groups[tn]
        group_info.append((tn, holes[0], holes[-1]))

    if len(tool_numbers) <= 7:
        best_order = tool_numbers
        best_dist = float("inf")
        for perm in permutations(range(len(group_info))):
            d = 0.0
            current = (0.0, 0.0)
            for idx in perm:
                entry = group_info[idx][1]
                d += _dist(current, entry)
                current = (group_info[idx][2].x, group_info[idx][2].y)
            if d < best_dist:
                best_dist = d
                best_order = [group_info[i][0] for i in perm]
        return best_order

    # NN for many tools
    visited = set()
    order = []
    current = (0.0, 0.0)
    for _ in group_info:
        best_d = float("inf")
        best_i = -1
        for j, (_, entry, _) in enumerate(group_info):
            if j not in visited:
                d = _dist(current, entry)
                if d < best_d:
                    best_d = d
                    best_i = j
        visited.add(best_i)
        order.append(group_info[best_i][0])
        current = (group_info[best_i][2].x, group_info[best_i][2].y)

    return order

File: drill_optimizer/js/drill_optimizer.js
/**
 * PCB Drill Path Optimizer
 *
 * Optimizes the order of drilling holes to minimize total travel distance.
 * Uses Nearest Neighbor construction + 2-opt local search improvement.
 * Groups holes by tool to minimize tool changes.
 */

const DrillOptimizer = (() => {

  /**
   * Calculate Euclidean distance between two points.
   */
  function dist(a, b) {
    const dx = a.x - b.x;
    const dy = a.y - b.y;
    return Math.sqrt(dx * dx + dy * dy);
  }

  /**
   * Calculate total path distance for a route.
   */
  function routeDistance(points, route, start) {
    let total = 0;
    let current = start || { x: 0, y: 0 };
    for (const idx of route) {
      total += dist(current, points[idx]);
      current = points[idx];
    }
    return total;
  }

  /**
   * Nearest Neighbor greedy TSP construction.
   * Always moves to the closest unvisited hole.
   * @param {Array<{x,y}>} points - hole coordinates
   * @param {{x,y}} start - starting position (default: origin)
   * @returns {number[]} - ordered indices
   */
  function nearestNeighbor(points, start) {
    const n = points.length;
    if (n === 0) return [];
    if (n === 1) return [0];

    const visited = new Array(n).fill(false);
    const route = [];
    let current = start || { x: 0, y: 0 };

    for (let step = 0; step < n; step++) {
      let bestDist = Infinity;
      let bestIdx = -1;

      for (let j = 0; j < n; j++) {
        if (!visited[j]) {
          const d = dist(current, points[j]);
          if (d < bestDist) {
            bestDist = d;
            bestIdx = j;
          }
        }
      }

      visited[bestIdx] = true;
      route.push(bestIdx);
      current = points[bestIdx];
    }

    return route;
  }

  /**
   * 2-opt local search improvement.
   * Repeatedly swaps two edges in the route if it reduces total distance.
   * @param {Array<{x,y}>} points - hole coordinates
   * @param {number[]} route - initial route (indices)
   * @param {number} maxIterations - max improvement passes (default 200)
   * @param {number} timeLimitMs - time limit in ms (default 5000)
   * @returns {number[]} - improved route
   */
  function twoOptImprove(points, route, maxIterations, timeLimitMs) {
    maxIterations = maxIterations || 200;
    timeLimitMs = timeLimitMs || 5000;

    const n = route.length;
    if (n < 4) return route.slice(); // Can't improve routes shorter than 4

    const improved = route.slice();
    const startTime = performance.now();
    let iterations = 0;

    // Pre-compute distance lookup for speed
    const distCache = new Float64Array(n * n);
    for (let i = 0; i < n; i++) {
      for (let j = i + 1; j < n; j++) {
        const d = dist(points[improved[i]], points[improved[j]]);
        distCache[i * n + j] = d;
        distCache[j * n + i] = d;
      }
    }

    let didImprove = true;
    while (didImprove && iterations < maxIterations) {
      didImprove = false;
      iterations++;

      if (performance.now() - startTime > timeLimitMs) break;

      for (let i = 1; i < n - 2; i++) {
        for (let j = i + 2; j < n; j++) {
          // Current edges: (i-1, i) and (j, j+1)
          // Proposed edges: (i-1, j) and (i, j+1)
          const a = improved[i - 1];
          const b = improved[i];
          const c = improved[j];
          const d = improved[j + 1 < n ? j + 1 : j];

          const currentDist = dist(points[a], points[b]) + dist(points[c], points[d]);
          const newDist = dist(points[a], points[c]) + dist(points[b], points[d]);

          if (newDist < currentDist - 1e-10) {
            // Reverse the segment between i and j
            const segment = improved.slice(i, j + 1);
            segment.reverse();
            for (let k = 0; k < segment.length; k++) {
              improved[i + k] = segment[k];
            }
            didImprove = true;
          }
        }
      }
    }

    return improved;
  }

  /**
   * Optimize drilling path for a complete drill project.
   * Groups holes by tool, optimizes each group independently,
   * then orders the tool groups to minimize inter-group travel.
   *
   * @param {Object} drillData - parsed Excellon data
   * @param {Object} options - { maxIterations: 200, timeLimitMs: 5000 }
   * @returns {Object} { orderedHoles, totalDistance, toolChanges, originalDistance, improvement, pathSegments }
   */
  function optimize(drillData, options) {
    options = options || {};
    const maxIter = options.maxIterations || 200;
    const timeLimit = options.timeLimitMs || 5000;

    // Group holes by tool
    const toolGroups = {};
    for (const hole of drillData.holes) {
      if (!toolGroups[hole.tool]) {
        toolGroups[hole.tool] = [];
      }
      toolGroups[hole.tool].push(hole);
    }

    const toolNumbers = Object.keys(toolGroups).map(Number).sort((a, b) => a - b);

    if (toolNumbers.length === 0) {
      return {
        orderedHoles: [],
        totalDistance: 0,
        toolChanges: 0,
        originalDistance: 0,
        improvement: 0,
        pathSegments: [],
        toolStats: []
      };
    }

    // Calculate original (unoptimized) distance for comparison
    const originalDistance = calculateOriginalDistance(drillData.holes);

    // Optimize each tool group
    const optimizedGroups = {};
    for (const toolNum of toolNumbers) {
      const holes = toolGroups[toolNum];
      const points = holes.map(h => ({ x: h.x, y: h.y }));

      // Nearest Neighbor from origin
      const nnRoute = nearestNeighbor(points, { x: 0, y: 0 });

      // 2-opt improvement
      const optRoute = twoOptImprove(points, nnRoute, maxIter, timeLimit);

      // Reorder holes according to optimized route
      optimizedGroups[toolNum] = optRoute.map(idx => holes[idx]);
    }

    // Order tool groups using nearest neighbor on group endpoints
    const groupOrder = orderToolGroups(optimizedGroups, toolNumbers);

    // Build final ordered hole list
    const orderedHoles = [];
    const toolChanges = [];
    let currentTool = null;

    for (const toolNum of groupOrder) {
      if (currentTool !== null && currentTool !== toolNum) {
        toolChanges.push({ from: currentTool, to: toolNum, index: orderedHoles.length });
      }
      currentTool = toolNum;
      for (const hole of optimizedGroups[toolNum]) {
        orderedHoles.push(hole);
      }
    }

    // Calculate optimized distance
    const totalDistance = calculatePathDistance(orderedHoles);

    // Build path segments for visualization
    const pathSegments = buildPathSegments(orderedHoles, toolChanges);

    // Per-tool statistics
    const toolStats = toolNumbers.map(tn => {
      const holes = optimizedGroups[tn];
      const toolDef = drillData.tools.find(t => t.number === tn);
      return {
        toolNumber: tn,
        diameter: toolDef ? toolDef.diameter : 0,
        hitCount: holes.length,
        subPathDistance: calculatePathDistance(holes)
      };
    });

    return {
      orderedHoles,
      totalDistance,
      toolChanges: toolChanges.length,
      originalDistance,
      improvement: originalDistance > 0 ? ((originalDistance - totalDistance) / originalDistance * 100) : 0,
      pathSegments,
      toolStats
    };
  }

  /**
   * Order tool groups to minimize inter-group travel distance.
   * Uses nearest-neighbor on the small TSP of group endpoints.
   */
  function orderToolGroups(groups, toolNumbers) {
    if (toolNumbers.length <= 1) return toolNumbers;

    // For each group, compute entry point (first hole) and exit point (last hole)
    const groupInfo = toolNumbers.map(tn => ({
      tool: tn,
      entry: groups[tn][0],
      exit: groups[tn][groups[tn].length - 1]
    }));

    // Try all permutations if few tools (<=7), otherwise use NN
    if (toolNumbers.length <= 7) {
      return bruteForceGroupOrder(groupInfo);
    }

    // Nearest neighbor on group entry points
    const visited = new Set();
    const order = [];
    let current = { x: 0, y: 0 };

    for (let step = 0; step < groupInfo.length; step++) {
      let bestDist = Infinity;
      let bestIdx = -1;

      for (let j = 0; j < groupInfo.length; j++) {
        if (!visited.has(j)) {
          const d = dist(current, groupInfo[j].entry);
          if (d < bestDist) {
            bestDist = d;
            bestIdx = j;
          }
        }
      }

      visited.add(bestIdx);
      order.push(groupInfo[bestIdx].tool);
      current = groupInfo[bestIdx].exit;
    }

    return order;
  }

  /**
   * Brute-force optimal ordering for small number of tool groups.
   */
  function bruteForceGroupOrder(groupInfo) {
    const n = groupInfo.length;
    let bestOrder = groupInfo.map(g => g.tool);
    let bestDist = Infinity;

    function permute(arr, start) {
      if (start === arr.length - 1) {
        let d = 0;
        let current = { x: 0, y: 0 };
        for (const idx of arr) {
          d += dist(current, groupInfo[idx].entry);
          current = groupInfo[idx].exit;
        }
        if (d < bestDist) {
          bestDist = d;
          bestOrder = arr.map(idx => groupInfo[idx].tool);
        }
        return;
      }
      for (let i = start; i < arr.length; i++) {
        [arr[start], arr[i]] = [arr[i], arr[start]];
        permute(arr, start + 1);
        [arr[start], arr[i]] = [arr[i], arr[start]];
      }
    }

    permute(groupInfo.map((_, i) => i), 0);
    return bestOrder;
  }

  /**
   * Calculate distance for holes in their original order.
   */
  function calculateOriginalDistance(holes) {
    let total = 0;
    let current = { x: 0, y: 0 };
    for (const hole of holes) {
      total += dist(current, hole);
      current = hole;
    }
    return total;
  }

  /**
   * Calculate distance for ordered holes.
   */
  function calculatePathDistance(holes) {
    return calculateOriginalDistance(holes);
  }

  /**
   * Build path segments for visualization.
   * Each segment: { from: {x,y}, to: {x,y}, type: 'rapid'|'drill'|'toolChange', tool }
   */
  function buildPathSegments(orderedHoles, toolChanges) {
    const segments = [];
    const changeIndices = new Set(toolChanges.map(tc => tc.index));
    let current = { x: 0, y: 0 };

    for (let i = 0; i < orderedHoles.length; i++) {
      const hole = orderedHoles[i];

      // Rapid move to hole position
      segments.push({
        from: { ...current },
        to: { x: hole.x, y: hole.y },
        type: changeIndices.has(i) ? 'toolChange' : 'rapid',
        tool: hole.tool
      });

      // Drill plunge (down and up - visualize as dot)
      segments.push({
        from: { x: hole.x, y: hole.y },
        to: { x: hole.x, y: hole.y },
        type: 'drill',
        tool: hole.tool
      });

      current = { x: hole.x, y: hole.y };
    }

    return segments;
  }

  return {
    optimize,
    nearestNeighbor,
    twoOptImprove,
    routeDistance
  };
})();
